### PR TITLE
Move proxy configuration form out of pluginManager screens as it is not related

### PR DESCRIPTION
**Pull request by olamy**
Signed-off-by: Olivier Lamy <olamy@apache.org>

See [JENKINS-72326](https://issues.jenkins.io/browse/JENKINS-72326).

### Proposed changelog entries

- Move the proxy configuration form to its own screen.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/ProxyConfigurationManager.java
Comment (daniel-beck) [inline core/src/main/java/hudson/ProxyConfigurationManager.java:40]: What is being managed here? Why not just
```suggestion
public class ProxyConfiguration extends ManagementLink {
```

File: core/src/main/java/hudson/ProxyConfigurationManager.java
Comment (daniel-beck) [inline core/src/main/java/hudson/ProxyConfigurationManager.java:39]:
```suggestion
@Extension @Restricted(NoExternalUse.class)
```

Plus imports

**Review by daniel-beck — COMMENTED** [review]
While it's a good idea to move the proxy configuration out of the plugin manager, what motivates it getting its own top-level `ManagementLink`? I would not expect that to be relevant regularly enough to justify the space it'll take up. Why not make it just another option in Configure System?

File: core/src/main/java/hudson/ProxyConfigurationManager.java
Comment (jtnord) [inline core/src/main/java/hudson/ProxyConfigurationManager.java:82]: I think this should be `Jenkins.MANAGE` as changing a proxy should not allow escallation of permissions to `Administer`?

File: core/src/main/resources/hudson/ProxyConfigurationManager/index.jelly
Comment (jtnord) [inline core/src/main/resources/hudson/ProxyConfigurationManager/index.jelly:29]: does this not need to set `readOnly` mode so the fields show as disabled when the user misses the permission to change them?

```suggestion

<j:set var="readOnlyMode" value="${!app.hasPermission(app.MANAGE)}"/>
```

**Comment by jtnord** [comment]
Is there not something in the setup wizard that lets you setup a proxy if it detects you have no connection, does that not need some changes to call the changed API?
https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/jenkins/install/SetupWizard/proxy-configuration.jelly

File: war/src/main/resources/images/symbols/proxy.svg
Comment (mawinter69) [inline war/src/main/resources/images/symbols/proxy.svg:1]: The image looks more like being related to WiFi than a proxy. This here https://icon-sets.iconify.design/eos-icons/proxy-outlined/ would be good but it's only 24x24

**Comment by olamy** [comment]
> While it's a good idea to move the proxy configuration out of the plugin manager, what motivates it getting its own top-level `ManagementLink`? I would not expect that to be relevant regularly enough to justify the space it'll take up. Why not make it just another option in Configure System?

Because I find the new icon really cute!

Sounds a good idea to add another option in Configure System. I will change to this.

File: core/src/main/java/hudson/ProxyConfigurationManager.java
Comment (olamy) [reply core/src/main/java/hudson/ProxyConfigurationManager.java:40]: uhm not sure the current class `ProxyConfiguration` from the same package will be happy with that :)

File: core/src/main/java/hudson/ProxyConfigurationManager.java
Comment (olamy) [reply core/src/main/java/hudson/ProxyConfigurationManager.java:82]: this is a copy of the current code. I didn't change that. (https://github.com/jenkinsci/jenkins/blob/51c5fa2367592f6fb6810f5eed89f88098ef0cf5/core/src/main/java/hudson/PluginManager.java#L1803)
Should it be change?

File: war/src/main/resources/images/symbols/proxy.svg
Comment (olamy) [reply war/src/main/resources/images/symbols/proxy.svg:1]: thanks for the research. But we might not need that as moving this change to Configure System screen.

**Comment by olamy** [comment]
> While it's a good idea to move the proxy configuration out of the plugin manager, what motivates it getting its own top-level `ManagementLink`? I would not expect that to be relevant regularly enough to justify the space it'll take up. Why not make it just another option in Configure System?


@daniel-beck 
done with 37918b91eded4d86337656c5e5d28db3dde04673

File: core/src/main/java/hudson/ProxyConfigurationManager.java
Comment (olamy) [reply core/src/main/java/hudson/ProxyConfigurationManager.java]: 0c0952862e40266d3d8add0693ad1b8f7c8d6bef

File: core/src/main/resources/hudson/ProxyConfigurationManager/index.jelly
Comment (olamy) [reply core/src/main/resources/hudson/ProxyConfigurationManager/index.jelly]: this mean modifying all config.groovy for ProxyConfiguration. This PR is only about moving the form not really changing how it works

File: core/src/main/java/hudson/ProxyConfigurationManager.java
Comment (olamy) [inline core/src/main/java/hudson/ProxyConfigurationManager.java:61]: I wonder if I should move this to the class ProxyConfiguration.
I'm tempted but so no strong opinion really

**Comment by olamy** [comment]
> Is there not something in the setup wizard that lets you setup a proxy if it detects you have no connection, does that not need some changes to call the changed API?
> 
> https://github.com/jenkinsci/jenkins/blob/92aa49868be12cc05cc28244b89eda900b68f7e9/core/src/main/resources/jenkins/install/SetupWizard/proxy-configuration.jelly#L14-L23

The same API will be still here.

**Review by prakash13594 — APPROVED** [review]

**Comment by prakash13594** [comment]
Changes Updated Proceed

File: core/src/main/java/hudson/ProxyConfigurationManager.java
Comment (mawinter69) [inline core/src/main/java/hudson/ProxyConfigurationManager.java:1]: How about naming this ProxyGlobalConfiguration?

**Review by mawinter69 — COMMENTED** [review]
Do we know if this affects CasC? I know there is special handling for proxy [here](https://github.com/jenkinsci/configuration-as-code-plugin/blob/097dcd0951febe91f3d4e468e9394bfdde3d0f87/plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java#L96) in CasC

**Comment by Kevin-CB** [comment]
Just to let you know, I will try to do a security review of this PR soonish

File: core/src/main/java/hudson/ProxyConfigurationManager.java
Comment (olamy) [reply core/src/main/java/hudson/ProxyConfigurationManager.java:1]: There is a class called `ProxyConfiguration` which already some contains some logic about this. 
So the idea was to have a name similar.
Anywat naming it's like color or codestyle or wine. Everybody may have his own taste preferences.
why not another name but someone may not like it or prefer something different, 
This will eventually turn into an endless loop and definitely a waste of time.

File: core/src/main/resources/hudson/PluginManager/advanced.jelly
Comment (timja) [inline core/src/main/resources/hudson/PluginManager/advanced.jelly:41]: do we want to put a notice here for a period of time to say it's moved?

**Review by timja — APPROVED** [review]
Thanks for this PR.

Two questions (non-blocking):

- The form is quite large, which was okay on it's own page but now on an instance with no plugins it takes up nearly a third of the system configuration page, is it worth hiding it behind a button like the administrative monitors?
- (inline)

File: core/src/main/resources/hudson/PluginManager/advanced.jelly
Comment (daniel-beck) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:41]: 👍 from me. I did that when we moved cloud configuration out of the Configure System page and thought that was nice (then we kept it around for too long but that's a separate issue).

Comment (olamy) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:41]: sure why not notice but what sort of style/color etc.. (fancy icon? :))

Comment (timja) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:41]: just the `-info` notification banner I think, (same type as admin monitors have but with just -info)

Comment (olamy) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:41]: done.

**Comment by timja** [comment]
> > * The form is quite large, which was okay on it's own page but now on an instance with no plugins it takes up nearly a third of the system configuration page, is it worth hiding it behind a button like the administrative monitors?
> 
> @timja
> 
> doesn't bother me too much compared to what you get when you have plenty of plugins installed :)
> 
> I'm not quite sure to follow " like the administrative monitors" where is the code for this?

On the system configuration page:

image

image

**Review by Kevin-CB — COMMENTED** [review]
Apologies for the delay. I've reviewed it and from a security standpoint, it looks fine.

File: core/src/main/resources/hudson/PluginManager/advanced.jelly
Comment (timja) [inline core/src/main/resources/hudson/PluginManager/advanced.jelly:41]: this breaks localisation as 'Configure System page' can't be localised. I think you want something like this and moving the link into property bundle:
```suggestion
              <strong>${%proxyMovedBlurb("${rootURL}/manage/configure")}</strong>
```

**Comment by jtnord** [comment]
> While it's a good idea to move the proxy configuration out of the plugin manager, what motivates it getting its own top-level ManagementLink? I would not expect that to be relevant regularly enough to justify the space it'll take up. Why not make it just another option in Configure System?

> the form is quite large, which was okay on it's own page but now on an instance with no plugins it takes up nearly a third of the system configuration page, is it worth hiding it behind a button like the administrative monitors?


the proxy is unusual as it does not configure Jenkins like the rest of the things in the `System` page, but really the JVM (it doesn't but its more a JVM thing than a Jenkins thing) which is why in my initial discussion with Olamy I suggested it was there.

You will likely need to configure the proxy to be able to other parts of the Jenkins config (like test connections etc).  Whilst you can "apply" or do a 2 step "save" this could also be needed for some security realms hence I still believe this is better as a separate item.

And then we have the discoverability.  if it has its own icon it is obvious - it was burried in the plugin section for a while, if we move it and hide it behind a button for users with a few of plugins installed will they easily be able to discover the new location?

I'm not going to die on this hill - but it seems like putting it into the the System page creates its own issues.

File: core/src/main/resources/hudson/PluginManager/advanced.jelly
Comment (jtnord) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:41]: this is not the configure system page but the plugin manager?
if this does not work then is the [other](https://github.com/jenkinsci/jenkins/blob/310bd6ac11939f525d055193b6d5d91ddc6ffcef/core/src/main/resources/hudson/PluginManager/advanced.jelly#L51) [localisation](https://github.com/jenkinsci/jenkins/blob/310bd6ac11939f525d055193b6d5d91ddc6ffcef/core/src/main/resources/hudson/PluginManager/advanced.jelly#L56) [in](https://github.com/jenkinsci/jenkins/blob/310bd6ac11939f525d055193b6d5d91ddc6ffcef/core/src/main/resources/hudson/PluginManager/advanced.jelly#L61) this page also broken?

**Comment by timja** [comment]
> And then we have the discoverability. if it has its own icon it is obvious - it was burried in the plugin section for a while, if we move it and hide it behind a button for users with a few of plugins installed will they easily be able to discover the new location?

I would expect most users would ctrl+F proxy, whether or not it's behind a button I don't think matters much.

----

I don't think proxy configuration is important enough or distinct enough to get it's own page.
It's in the setup wizard anyway so should be setup in first time setup first.

File: core/src/main/resources/hudson/PluginManager/advanced.jelly
Comment (olamy) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly]: fixed with 48d3fcf691

**Comment by olamy** [comment]
@timja use some advanced done 8315dd5cf3

**Review by timja — APPROVED** [review]
Thanks, re-tested looks good!

**Comment by daniel-beck** [comment]
> discoverability

I don't think this is any different from any other option in Configure System? The same argument can be made for every option there.

> it was burried in the plugin section for a while, if we move it and hide it behind a button for users with a few of plugins installed will they easily be able to discover the new location?

Yes, if we do it like we did for the cloud config (leaving behind a placeholder, https://github.com/jenkinsci/jenkins/pull/4339 third screenshot).

> You will likely need to configure the proxy to be able to other parts of the Jenkins config (like test connections etc).

This argument makes more sense to me. OTOH at least the form allows for it, while otherwise system config needs to be abandoned to set up the proxy first.

Note that

> could also be needed for some security realms

is less relevant, security realms are on Configure Global Security, so separate from either. Unless you're saying admins changing security options cannot be expected to visit Configure System beforehand if their environment demands it?

**Review by jtnord — APPROVED** [review]

File: core/src/main/java/hudson/ProxyConfigurationManager.java
Comment (jtnord) [reply core/src/main/java/hudson/ProxyConfigurationManager.java]: > Should it be change?

let's not do it here - can be done in a follow up as it may require some thought as to if it can be abused (I do not see a vector to escalate here - but at least it would require a security review)

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
